// @plasmicpkgs/plasmic-rich-components/skinny/common-prop-types-988c65db.esm.js
import { c as getFieldSubprops, b as buildFieldsPropType } from './formatting-0acc57f5.esm.js';
import { c as capitalize } from './common-8cca2977.esm.js';
import { m as maybe } from './utils-c32bd7ed.esm.js';

function roleProp({
  role,
  singular = false,
  advanced = false,
  displayName
}) {
  return singular ? {
    type: "object",
    displayName: displayName || `${capitalize(role)} field`,
    advanced,
    hidden: (ps) => !ps.data,
    nameFunc: (item) => maybe(
      item,
      (i) => i.isHidden ? "Hidden" : i.fieldId || "Custom value"
    ),
    fields: getFieldSubprops({
      canChangeField: true,
      noTitle: true
    }),
    defaultValueHint: (_props, contextData) => {
      var _a;
      return ((_a = contextData == null ? void 0 : contextData.minimalFullLengthFields) != null ? _a : []).find(
        (f) => f.role === role
      );
    }
  } : buildFieldsPropType({
    displayName: displayName || `${capitalize(role)} fields`,
    advanced,
    noTitle: true,
    canChangeField: true,
    minimalValue: (_props, contextData) => {
      var _a;
      return ((_a = contextData == null ? void 0 : contextData.minimalFullLengthFields) != null ? _a : []).filter(
        (f) => f.role === role
      );
    }
  });
}
function dataProp() {
  return {
    type: "dataSourceOpData",
    description: "The data to display"
  };
}
function commonProps() {
  return {
    pagination: {
      type: "boolean",
      advanced: true,
      defaultValueHint: true
    },
    pageSize: {
      type: "number",
      defaultValueHint: 10,
      advanced: true
    },
    hideSearch: {
      type: "boolean",
      description: "Hides the search toolbar",
      advanced: true
    }
  };
}
function rowActionsProp() {
  return {
    type: "array",
    displayName: "Row actions",
    advanced: true,
    itemType: {
      type: "object",
      nameFunc: (item) => item.label,
      fields: {
        type: {
          type: "choice",
          options: ["item", "menu"],
          defaultValue: "item"
        },
        label: {
          type: "string",
          displayName: "Action label"
        },
        children: {
          type: "array",
          displayName: "Menu items",
          itemType: {
            type: "object",
            fields: {
              label: {
                type: "string",
                displayName: "Action label"
              },
              onClick: {
                type: "eventHandler",
                argTypes: [
                  { name: "rowKey", type: "string" },
                  { name: "row", type: "object" }
                ]
              }
            }
          },
          hidden: (_ps, _ctx, { item }) => item.type !== "menu"
        },
        onClick: {
          type: "eventHandler",
          displayName: "Action",
          argTypes: [
            { name: "rowKey", type: "string" },
            { name: "row", type: "object" }
          ],
          hidden: (_ps, _ctx, { item }) => item.type !== "item"
        }
      }
    }
  };
}
function onRowClickProp() {
  return {
    type: "eventHandler",
    displayName: "On row clicked",
    argTypes: [
      { name: "rowKey", type: "string" },
      { name: "row", type: "object" },
      { name: "event", type: "object" }
    ]
  };
}

export { rowActionsProp as a, commonProps as c, dataProp as d, onRowClickProp as o, roleProp as r };
