// @plasmicpkgs/plasmic-rich-components/skinny/rich-details/index.esm.js
import { r as renderValue, b as buildFieldsPropType } from '../formatting-0acc57f5.esm.js';
import { d as mkShortId, r as registerComponentHelper } from '../utils-c32bd7ed.esm.js';
import { useNormalizedData, deriveFieldConfigs } from '@plasmicapp/data-sources';
import { Empty, Descriptions } from 'antd';
import React from 'react';
import '@plasmicpkgs/luxon-parser';
import 'lodash/get';
import '@plasmicapp/host/registerComponent';
import '@plasmicapp/host/registerGlobalContext';
import 'dayjs';
import 'dayjs/plugin/customParseFormat';

var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
function RichDetails(props) {
  var _a;
  const {
    className,
    data: rawData,
    size,
    bordered,
    layout,
    column = 2
  } = props;
  const data = useNormalizedData(rawData);
  const { columnDefinitions } = useColumnDefinitions(data, props);
  if (!data || !((_a = data.data) == null ? void 0 : _a[0])) {
    return /* @__PURE__ */ React.createElement(Empty, { className, image: Empty.PRESENTED_IMAGE_SIMPLE });
  }
  const row = data.data[0];
  return /* @__PURE__ */ React.createElement(
    Descriptions,
    {
      className,
      size,
      bordered,
      layout,
      column: {
        xs: 1,
        sm: 1,
        md: column
      }
    },
    columnDefinitions.map((col) => /* @__PURE__ */ React.createElement(Descriptions.Item, { label: col.title, key: col.key, span: col.span }, col.render(row)))
  );
}
function useColumnDefinitions(data, props) {
  const { fields, setControlContextData } = props;
  return React.useMemo(() => {
    const schema = data == null ? void 0 : data.schema;
    if (!data || !schema) {
      return { normalized: [], columnDefinitions: [] };
    }
    const { mergedFields, minimalFullLengthFields } = deriveFieldConfigs(
      fields != null ? fields : [],
      schema,
      (field) => __spreadValues({
        key: mkShortId(),
        isHidden: false,
        dataType: "auto"
      }, field && {
        key: field.id,
        fieldId: field.id,
        title: field.label || field.id,
        expr: (currentItem) => currentItem[field.id]
      })
    );
    setControlContextData == null ? void 0 : setControlContextData(__spreadProps(__spreadValues({}, data), { mergedFields, minimalFullLengthFields }));
    const normalized = mergedFields;
    const columnDefinitions = normalized.filter((cconfig) => !cconfig.isHidden).map((cconfig, _columnIndex, _columnsArray) => {
      const columnDefinition = {
        dataIndex: cconfig.fieldId,
        title: cconfig.title,
        key: cconfig.key,
        span: cconfig.span,
        render: (record) => {
          return renderValue(record, cconfig);
        }
      };
      return columnDefinition;
    });
    return { normalized, columnDefinitions };
  }, [fields, data, setControlContextData]);
}

function registerRichDetails(loader) {
  registerComponentHelper(loader, RichDetails, {
    name: "hostless-rich-details",
    displayName: "Data details",
    defaultStyles: {
      width: "stretch",
      maxHeight: "100%"
    },
    props: {
      data: {
        type: "dataSourceOpData",
        description: "The data to display"
      },
      fields: buildFieldsPropType({
        fieldTypes: {
          span: {
            type: "number",
            displayName: "Number of columns to span",
            defaultValueHint: 1
          }
        }
      }),
      layout: {
        displayName: "Layout",
        type: "choice",
        options: [
          { value: "horizontal", label: "Horizontal" },
          { value: "vertical", label: "Vertical" }
        ],
        defaultValueHint: "horizontal"
      },
      column: {
        displayName: "Items per row",
        type: "number",
        description: "Number of items to display per row",
        defaultValueHint: 2
      },
      size: {
        displayName: "Spacing",
        type: "choice",
        options: [
          { value: "small", label: "Small" },
          { value: "middle", label: "Medium" },
          { value: "default", label: "Large" }
        ],
        defaultValueHint: "default"
      },
      bordered: {
        displayName: "Show borders?",
        type: "boolean",
        defaultValue: true
      }
    },
    importPath: "@plasmicpkgs/plasmic-rich-components/skinny/rich-details",
    importName: "RichDetails"
  });
}

export { RichDetails, registerRichDetails };
