// @plasmicpkgs/react-aria/skinny/registerTextArea.cjs.js
'use strict';

var React = require('react');
var reactAria = require('react-aria');
var reactAriaComponents = require('react-aria-components');
var common = require('./common-6ffc5113.cjs.js');
var contexts = require('./contexts-6d0cb2b1.cjs.js');
var utils = require('./utils-88dd1a55.cjs.js');
var variantUtils = require('./variant-utils-0ad70db8.cjs.js');
require('@plasmicapp/host');
require('@plasmicapp/host/registerComponent');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var React__default = /*#__PURE__*/_interopDefault(React);

var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
const TEXTAREA_VARIANTS = [
  "focused",
  "focusVisible",
  "hovered",
  "disabled"
];
const { variants } = variantUtils.pickAriaComponentVariants(TEXTAREA_VARIANTS);
const inputHelpers = {
  states: {
    value: {
      onChangeArgsToValue: (e) => {
        return e.target.value;
      }
    }
  }
};
function BaseTextArea_(props, ref) {
  const _a = props, {
    className,
    plasmicUpdateVariant,
    setControlContextData,
    value
  } = _a, restProps = __objRest(_a, [
    "className",
    "plasmicUpdateVariant",
    "setControlContextData",
    "value"
  ]);
  const textFieldContext = React__default.default.useContext(contexts.PlasmicTextFieldContext);
  const [textAreaContextProps, textAreaRef] = reactAriaComponents.useContextProps(
    restProps,
    ref,
    reactAriaComponents.TextAreaContext
  );
  const { hoverProps, isHovered } = reactAria.useHover(textAreaContextProps);
  const { isFocused, isFocusVisible, focusProps } = reactAria.useFocusRing({
    isTextInput: true,
    autoFocus: textAreaContextProps.autoFocus
  });
  const mergedProps = reactAria.mergeProps(
    utils.filterHoverProps(textAreaContextProps),
    focusProps,
    hoverProps,
    {
      value: utils.isDefined(textFieldContext) ? void 0 : value,
      style: common.COMMON_STYLES,
      className
    }
  );
  const isDisabled = mergedProps.disabled || false;
  const isInvalid = !!mergedProps["aria-invalid"] && mergedProps["aria-invalid"] !== "false";
  setControlContextData == null ? void 0 : setControlContextData({
    parent: textFieldContext
  });
  React.useEffect(() => {
    if (plasmicUpdateVariant) {
      plasmicUpdateVariant({
        disabled: isDisabled,
        focused: isFocused,
        focusVisible: isFocusVisible,
        hovered: isHovered
      });
    }
  }, [isFocused, isFocusVisible, isHovered, isDisabled, plasmicUpdateVariant]);
  return /* @__PURE__ */ React__default.default.createElement(
    "textarea",
    __spreadProps(__spreadValues({}, mergedProps), {
      ref: textAreaRef,
      "data-focused": isFocused || void 0,
      "data-disabled": isDisabled || void 0,
      "data-hovered": isHovered || void 0,
      "data-focus-visible": isFocusVisible || void 0,
      "data-invalid": isInvalid || void 0
    })
  );
}
const BaseTextArea = React__default.default.forwardRef(BaseTextArea_);
function registerTextArea(loader, overrides) {
  utils.registerComponentHelper(
    loader,
    BaseTextArea,
    {
      name: utils.makeComponentName("textarea"),
      displayName: "Aria TextArea",
      importPath: "@plasmicpkgs/react-aria/skinny/registerTextArea",
      importName: "BaseTextArea",
      variants,
      props: __spreadValues({}, common.getCommonProps("Text Area", [
        "name",
        "disabled",
        "readOnly",
        "autoFocus",
        "aria-label",
        "required",
        "placeholder",
        "value",
        "maxLength",
        "minLength",
        "inputMode",
        "onChange",
        "onFocus",
        "onBlur",
        "onKeyDown",
        "onKeyUp",
        "onCopy",
        "onCut",
        "onPaste",
        "onCompositionStart",
        "onCompositionEnd",
        "onCompositionUpdate",
        "onSelect",
        "onBeforeInput",
        "onInput"
      ])),
      states: {
        value: __spreadValues({
          type: "writable",
          valueProp: "value",
          onChangeProp: "onChange",
          variableType: "text"
        }, inputHelpers.states.value)
      },
      componentHelpers: {
        helpers: inputHelpers,
        importName: "inputHelpers",
        importPath: "@plasmicpkgs/react-aria/skinny/registerTextArea"
      }
    },
    overrides
  );
}

exports.BaseTextArea = BaseTextArea;
exports.inputHelpers = inputHelpers;
exports.registerTextArea = registerTextArea;
