// @plasmicpkgs/react-aria/skinny/registerInput.esm.js
import React, { useEffect } from 'react';
import { useHover, useFocusRing, mergeProps } from 'react-aria';
import { useContextProps, InputContext } from 'react-aria-components';
import { C as COMMON_STYLES, r as resolveAutoComplete, g as getCommonProps } from './common-3fde23cc.esm.js';
import { g as PlasmicTextFieldContext, c as PlasmicInputContext } from './contexts-5cb81c2f.esm.js';
import { f as filterHoverProps, i as isDefined, m as makeComponentName, r as registerComponentHelper } from './utils-b68e14cf.esm.js';
import { p as pickAriaComponentVariants } from './variant-utils-4405ebb0.esm.js';
import '@plasmicapp/host';
import '@plasmicapp/host/registerComponent';

var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
const INPUT_VARIANTS = [
  "focused",
  "focusVisible",
  "hovered",
  "disabled"
];
const { variants } = pickAriaComponentVariants(INPUT_VARIANTS);
const inputHelpers = {
  states: {
    value: {
      onChangeArgsToValue: (e) => {
        return e.target.value;
      }
    }
  }
};
function BaseInput_(props, ref) {
  const _a = props, {
    plasmicUpdateVariant,
    setControlContextData,
    autoComplete,
    value,
    className
  } = _a, restProps = __objRest(_a, [
    "plasmicUpdateVariant",
    "setControlContextData",
    "autoComplete",
    "value",
    "className"
  ]);
  const textFieldContext = React.useContext(PlasmicTextFieldContext);
  const context = React.useContext(PlasmicInputContext);
  const [inputContextProps, inputRef] = useContextProps(
    restProps,
    ref,
    InputContext
  );
  const { hoverProps, isHovered } = useHover(props);
  const { isFocused, isFocusVisible, focusProps } = useFocusRing({
    isTextInput: true,
    autoFocus: inputContextProps.autoFocus
  });
  const mergedProps = mergeProps(
    filterHoverProps(inputContextProps),
    focusProps,
    hoverProps,
    {
      style: COMMON_STYLES,
      value: (context == null ? void 0 : context.isUncontrolled) || isDefined(textFieldContext) ? void 0 : value,
      autoComplete: resolveAutoComplete(autoComplete),
      className
    }
  );
  const isDisabled = mergedProps.disabled || false;
  const isInvalid = !!mergedProps["aria-invalid"] && mergedProps["aria-invalid"] !== "false";
  setControlContextData == null ? void 0 : setControlContextData({
    parent: textFieldContext
  });
  useEffect(() => {
    if (plasmicUpdateVariant) {
      plasmicUpdateVariant({
        disabled: isDisabled,
        focused: isFocused,
        focusVisible: isFocusVisible,
        hovered: isHovered
      });
    }
  }, [isFocused, isFocusVisible, isHovered, isDisabled, plasmicUpdateVariant]);
  return /* @__PURE__ */ React.createElement(
    "input",
    __spreadProps(__spreadValues({}, mergedProps), {
      ref: inputRef,
      "data-focused": isFocused || void 0,
      "data-disabled": isDisabled || void 0,
      "data-hovered": isHovered || void 0,
      "data-focus-visible": isFocusVisible || void 0,
      "data-invalid": isInvalid || void 0
    })
  );
}
const BaseInput = React.forwardRef(BaseInput_);
const INPUT_COMPONENT_NAME = makeComponentName("input");
function registerInput(loader, overrides) {
  registerComponentHelper(
    loader,
    BaseInput,
    {
      name: INPUT_COMPONENT_NAME,
      displayName: "Aria Input",
      importPath: "@plasmicpkgs/react-aria/skinny/registerInput",
      importName: "BaseInput",
      variants,
      defaultStyles: {
        width: "300px",
        borderWidth: "1px",
        borderStyle: "solid",
        borderColor: "black",
        padding: "4px 10px"
      },
      props: __spreadValues({}, getCommonProps("Input", [
        "name",
        "disabled",
        "readOnly",
        "autoFocus",
        "aria-label",
        "required",
        "placeholder",
        "value",
        "maxLength",
        "minLength",
        "pattern",
        "type",
        "inputMode",
        "autoComplete",
        "onChange",
        "onFocus",
        "onBlur",
        "onKeyDown",
        "onKeyUp",
        "onCopy",
        "onCut",
        "onPaste",
        "onCompositionStart",
        "onCompositionEnd",
        "onCompositionUpdate",
        "onSelect",
        "onBeforeInput",
        "onInput"
      ])),
      states: {
        value: __spreadValues({
          type: "writable",
          valueProp: "value",
          onChangeProp: "onChange",
          variableType: "text"
        }, inputHelpers.states.value)
      },
      componentHelpers: {
        helpers: inputHelpers,
        importName: "inputHelpers",
        importPath: "@plasmicpkgs/react-aria/skinny/registerInput"
      }
    },
    overrides
  );
}

export { BaseInput, INPUT_COMPONENT_NAME, inputHelpers, registerInput };
