// @plasmicpkgs/antd5/skinny/Form.cjs.js
'use strict';

var antd = require('antd');
var equal = require('fast-deep-equal');
var React = require('react');
var utils = require('./utils-7711f93b.cjs.js');
var contexts = require('./contexts-a2387bed.cjs.js');
require('@plasmicapp/host/registerComponent');
require('@plasmicapp/host/registerGlobalContext');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var equal__default = /*#__PURE__*/_interopDefault(equal);
var React__default = /*#__PURE__*/_interopDefault(React);

var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
var InputType = /* @__PURE__ */ ((InputType2) => {
  InputType2["Text"] = "Text";
  InputType2["TextArea"] = "Text Area";
  InputType2["Password"] = "Password";
  InputType2["Number"] = "Number";
  InputType2["Select"] = "Select";
  InputType2["Option"] = "Option";
  InputType2["OptionGroup"] = "Option Group";
  InputType2["Radio"] = "Radio";
  InputType2["RadioGroup"] = "Radio Group";
  InputType2["Checkbox"] = "Checkbox";
  InputType2["DatePicker"] = "DatePicker";
  InputType2["Unknown"] = "Unkown";
  return InputType2;
})(InputType || {});
const SchemaFormContext = React__default.default.createContext(void 0);
var FormType = /* @__PURE__ */ ((FormType2) => {
  FormType2[FormType2["NewEntry"] = 0] = "NewEntry";
  FormType2[FormType2["UpdateEntry"] = 1] = "UpdateEntry";
  return FormType2;
})(FormType || {});
const Internal = React__default.default.forwardRef(
  (props, ref) => {
    var _b, _c, _d, _e;
    const [isSubmitting, setIsSubmitting] = React__default.default.useState(false);
    const [form] = antd.Form.useForm();
    const values = form.getFieldsValue(true);
    const lastValue = React__default.default.useRef(values);
    const _a = props, {
      extendedOnValuesChange,
      forceRemount,
      formLayout,
      internalFieldCtx,
      setInternalFieldCtx,
      autoDisableWhileSubmitting = true
    } = _a, rest = __objRest(_a, [
      "extendedOnValuesChange",
      "forceRemount",
      "formLayout",
      "internalFieldCtx",
      "setInternalFieldCtx",
      "autoDisableWhileSubmitting"
    ]);
    const childrenNode = typeof props.children === "function" ? props.children(values, form) : props.children;
    const fireOnValuesChange = React__default.default.useCallback(() => {
      const values2 = form.getFieldsValue(true);
      if (!equal__default.default(values2, lastValue.current)) {
        extendedOnValuesChange == null ? void 0 : extendedOnValuesChange(values2);
        lastValue.current = values2;
      }
    }, [form, lastValue]);
    React__default.default.useEffect(() => {
      fireOnValuesChange();
    }, []);
    React__default.default.useImperativeHandle(ref, () => ({
      formInstance: form,
      setFieldsValue: (newValues) => {
        form.setFieldsValue(newValues);
        extendedOnValuesChange == null ? void 0 : extendedOnValuesChange(form.getFieldsValue(true));
      },
      setFieldValue: (namePath, value) => {
        form.setFieldValue(namePath, value);
        extendedOnValuesChange == null ? void 0 : extendedOnValuesChange(form.getFieldsValue(true));
      },
      resetFields: () => {
        form.resetFields();
        extendedOnValuesChange == null ? void 0 : extendedOnValuesChange(form.getFieldsValue(true));
      },
      validateFields: async (...args) => {
        try {
          return await form.validateFields(...args);
        } catch (err) {
          return err;
        }
      },
      clearFields: () => {
        const values2 = form.getFieldsValue(true);
        utils.setFieldsToUndefined(values2);
        form.setFieldsValue(values2);
        extendedOnValuesChange == null ? void 0 : extendedOnValuesChange(form.getFieldsValue(true));
      }
    }));
    const registerField = React__default.default.useCallback(
      (fieldEntity) => {
        setInternalFieldCtx((ctx) => ({
          registeredFields: [...ctx.registeredFields, fieldEntity],
          preservedRegisteredFields: [
            ...ctx.preservedRegisteredFields,
            fieldEntity
          ]
        }));
        return () => {
          setInternalFieldCtx((ctx) => ({
            registeredFields: ctx.registeredFields.filter(
              (ent) => ent !== fieldEntity
            ),
            preservedRegisteredFields: ctx.preservedRegisteredFields.filter(
              (ent) => ent !== fieldEntity || fieldEntity.preserve
            )
          }));
        };
      },
      [setInternalFieldCtx]
    );
    const schemaFormCtx = React__default.default.useContext(SchemaFormContext);
    (_b = props.setControlContextData) == null ? void 0 : _b.call(props, __spreadValues({
      formInstance: form,
      layout: formLayout,
      internalFieldCtx
    }, schemaFormCtx ? schemaFormCtx : {}));
    const updateIsSubmitting = React__default.default.useCallback(
      (newValue) => {
        var _a2;
        setIsSubmitting(newValue);
        (_a2 = props.onIsSubmittingChange) == null ? void 0 : _a2.call(props, newValue);
      },
      [props.onIsSubmittingChange, setIsSubmitting]
    );
    return /* @__PURE__ */ React__default.default.createElement(
      contexts.InternalFormInstanceContext.Provider,
      {
        value: {
          layout: formLayout,
          fireOnValuesChange,
          forceRemount,
          registerField,
          internalFieldCtx,
          initialValues: (_c = props.initialValues) != null ? _c : {}
        }
      },
      /* @__PURE__ */ React__default.default.createElement(contexts.FormLayoutContext.Provider, { value: formLayout }, /* @__PURE__ */ React__default.default.createElement(
        antd.Form,
        __spreadProps(__spreadValues({}, rest), {
          key: props.initialValues ? JSON.stringify(props.initialValues) : void 0,
          onValuesChange: (...args) => {
            var _a2;
            (_a2 = props.onValuesChange) == null ? void 0 : _a2.call(props, ...args);
            extendedOnValuesChange == null ? void 0 : extendedOnValuesChange(form.getFieldsValue(true));
          },
          onFinish: async () => {
            var _a2;
            if (isSubmitting && autoDisableWhileSubmitting) {
              return;
            }
            updateIsSubmitting(true);
            const submission = (_a2 = props.onFinish) == null ? void 0 : _a2.call(
              props,
              utils.pick(
                form.getFieldsValue(true),
                ...internalFieldCtx.preservedRegisteredFields.map(
                  (field) => field.fullPath
                )
              )
            );
            if (typeof submission === "object" && typeof submission.then === "function") {
              await submission;
            }
            updateIsSubmitting(false);
          },
          form,
          labelCol: ((_d = props.labelCol) == null ? void 0 : _d.horizontalOnly) && props.layout !== "horizontal" ? void 0 : props.labelCol,
          wrapperCol: ((_e = props.wrapperCol) == null ? void 0 : _e.horizontalOnly) && props.layout !== "horizontal" ? void 0 : props.wrapperCol,
          disabled: isSubmitting && autoDisableWhileSubmitting
        }),
        /* @__PURE__ */ React__default.default.createElement("style", null, `
          .ant-form-item-explain + div, .ant-form-item-margin-offset {
            display: none;
          }
          `),
        childrenNode
      ))
    );
  }
);
const FormWrapper = React__default.default.forwardRef(
  (props, ref) => {
    var _a;
    const [remountKey, setRemountKey] = React__default.default.useState(0);
    const forceRemount = React__default.default.useCallback(
      () => setRemountKey((k) => k + 1),
      [setRemountKey]
    );
    const previousInitialValues = utils.usePrevious(props.initialValues);
    const wrapperRef = React__default.default.useRef(null);
    React__default.default.useEffect(() => {
      if (previousInitialValues !== props.initialValues && JSON.stringify(previousInitialValues) !== JSON.stringify(props.initialValues)) {
        forceRemount();
      }
    }, [previousInitialValues, props.initialValues]);
    const [internalFieldCtx, setInternalFieldCtx] = React__default.default.useState({
      registeredFields: [],
      preservedRegisteredFields: []
    });
    React__default.default.useImperativeHandle(
      ref,
      () => wrapperRef.current ? __spreadValues({}, wrapperRef.current) : {}
    );
    const formLayout = React__default.default.useMemo(
      () => {
        var _a2;
        return {
          layout: props.layout,
          labelSpan: (_a2 = props.labelCol) == null ? void 0 : _a2.span
        };
      },
      [props.layout, (_a = props.labelCol) == null ? void 0 : _a.span]
    );
    return /* @__PURE__ */ React__default.default.createElement(
      Internal,
      __spreadValues({
        key: remountKey,
        forceRemount,
        formLayout,
        internalFieldCtx,
        setInternalFieldCtx,
        ref: wrapperRef
      }, props)
    );
  }
);
const formHelpers = {
  states: {
    value: {
      onMutate: (value, $ref) => {
        var _a;
        (_a = $ref == null ? void 0 : $ref.formInstance) == null ? void 0 : _a.setFieldsValue(value);
      }
    }
  }
};
const OPTIMIZED_FORM_IMPORT = {
  name: "FormWrapper",
  path: "@plasmicpkgs/antd5/skinny/Form"
};

exports.FormType = FormType;
exports.FormWrapper = FormWrapper;
exports.InputType = InputType;
exports.OPTIMIZED_FORM_IMPORT = OPTIMIZED_FORM_IMPORT;
exports.SchemaFormContext = SchemaFormContext;
exports.formHelpers = formHelpers;
