// @plasmicpkgs/antd5/skinny/names-73583568.esm.js
const inputComponentName = "plasmic-antd5-input";
const passwordComponentName = "plasmic-antd5-input-password";
const inputNumberComponentName = "plasmic-antd5-input-number";
const textAreaComponentName = "plasmic-antd5-textarea";
const switchComponentName = "plasmic-antd5-switch";
const selectComponentName = "plasmic-antd5-select";
const optionComponentName = "plasmic-antd5-option";
const optionGroupComponentName = "plasmic-antd5-option-group";
const checkboxComponentName = "plasmic-antd5-checkbox";
const checkboxGroupComponentName = "plasmic-antd5-checkbox-group";
const radioComponentName = "plasmic-antd5-radio";
const radioButtonComponentName = "plasmic-antd5-radio-button";
const radioGroupComponentName = "plasmic-antd5-radio-group";
const formComponentName = "plasmic-antd5-form";
const buttonComponentName = "plasmic-antd5-button";

export { checkboxGroupComponentName as a, buttonComponentName as b, checkboxComponentName as c, inputNumberComponentName as d, radioGroupComponentName as e, formComponentName as f, radioButtonComponentName as g, optionGroupComponentName as h, inputComponentName as i, switchComponentName as j, optionComponentName as o, passwordComponentName as p, radioComponentName as r, selectComponentName as s, textAreaComponentName as t };
