// @plasmicpkgs/antd5/skinny/names-3aea627b.cjs.js
'use strict';

const inputComponentName = "plasmic-antd5-input";
const passwordComponentName = "plasmic-antd5-input-password";
const inputNumberComponentName = "plasmic-antd5-input-number";
const textAreaComponentName = "plasmic-antd5-textarea";
const switchComponentName = "plasmic-antd5-switch";
const selectComponentName = "plasmic-antd5-select";
const optionComponentName = "plasmic-antd5-option";
const optionGroupComponentName = "plasmic-antd5-option-group";
const checkboxComponentName = "plasmic-antd5-checkbox";
const checkboxGroupComponentName = "plasmic-antd5-checkbox-group";
const radioComponentName = "plasmic-antd5-radio";
const radioButtonComponentName = "plasmic-antd5-radio-button";
const radioGroupComponentName = "plasmic-antd5-radio-group";
const formComponentName = "plasmic-antd5-form";
const buttonComponentName = "plasmic-antd5-button";

exports.buttonComponentName = buttonComponentName;
exports.checkboxComponentName = checkboxComponentName;
exports.checkboxGroupComponentName = checkboxGroupComponentName;
exports.formComponentName = formComponentName;
exports.inputComponentName = inputComponentName;
exports.inputNumberComponentName = inputNumberComponentName;
exports.optionComponentName = optionComponentName;
exports.optionGroupComponentName = optionGroupComponentName;
exports.passwordComponentName = passwordComponentName;
exports.radioButtonComponentName = radioButtonComponentName;
exports.radioComponentName = radioComponentName;
exports.radioGroupComponentName = radioGroupComponentName;
exports.selectComponentName = selectComponentName;
exports.switchComponentName = switchComponentName;
exports.textAreaComponentName = textAreaComponentName;
