// @plasmicpkgs/antd5/skinny/SchemaForm.esm.js
import { deriveFieldConfigs, normalizeData, usePlasmicDataOp } from '@plasmicapp/data-sources';
import { Input, InputNumber } from 'antd';
import React from 'react';
import { AntdCheckbox } from './registerCheckbox.esm.js';
import { AntdDatePicker } from './registerDatePicker.esm.js';
import { AntdRadioGroup } from './registerRadio.esm.js';
import { AntdSelect } from './registerSelect.esm.js';
import { u as usePrevious, E as ErrorBoundary, o as omit } from './utils-5ebcaa63.esm.js';
import { InputType, SchemaFormContext, FormWrapper } from './Form.esm.js';
import { FormItemWrapper } from './FormItem.esm.js';
import './names-73583568.esm.js';
import 'dayjs';
import './react-utils-79d444f6.esm.js';
import 'classnames';
import '@plasmicapp/host/registerComponent';
import '@plasmicapp/host/registerGlobalContext';
import 'fast-deep-equal';
import './contexts-4c7952c1.esm.js';
import '@plasmicapp/host';

var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
function deriveFormFieldConfigs(dataFormItems, schema, data) {
  return deriveFieldConfigs(
    dataFormItems,
    schema,
    (field) => {
      var _a;
      return __spreadValues({
        inputType: InputType.Text
      }, field && {
        key: field.id,
        fieldId: field.id,
        label: (_a = field.label) != null ? _a : field.id,
        name: field.id,
        inputType: field.type === "string" ? InputType.Text : field.type === "number" ? InputType.Number : field.type === "boolean" ? InputType.Checkbox : InputType.Text,
        //missing date and date-time
        initialValue: data ? data[field.id] : void 0
      });
    }
  );
}
function useFormItemDefinitions(rawData, props) {
  const { mode, dataFormItems, setControlContextData } = props;
  return React.useMemo(() => {
    const data = rawData && normalizeData(rawData);
    const schema = data && (data == null ? void 0 : data.schema);
    if (mode !== "simplified" || !rawData || rawData.isLoading || rawData.error || !data || !schema || !data.data) {
      return void 0;
    }
    const row = data.data.length > 0 ? data.data[0] : void 0;
    return deriveFormFieldConfigs(dataFormItems != null ? dataFormItems : [], schema, row);
  }, [mode, setControlContextData, dataFormItems, rawData]);
}
const useRawData = (props) => {
  const rawData = usePlasmicDataOp(props.data);
  return props.data ? rawData : void 0;
};
const SchemaForm = React.forwardRef(
  (props, ref) => {
    const [remountKey, setRemountKey] = React.useState(0);
    const forceRemount = React.useCallback(
      () => setRemountKey((k) => k + 1),
      [setRemountKey]
    );
    const wrapperRef = React.useRef(null);
    React.useImperativeHandle(
      ref,
      () => wrapperRef.current ? __spreadValues({}, wrapperRef.current) : {}
    );
    const rawData = useRawData(props);
    const formItemDefinitions = useFormItemDefinitions(rawData, props);
    React.useEffect(() => {
      if (rawData && !rawData.isLoading) {
        forceRemount();
      }
    }, [rawData]);
    const previousDataOp = usePrevious(props.data);
    React.useEffect(() => {
      if (previousDataOp == null && props.data != null || previousDataOp != null && props.data == null) {
        forceRemount();
      }
    }, [props.data]);
    const _a = props, { dataFormItems, formItems, data } = _a, rest = __objRest(_a, ["dataFormItems", "formItems", "data"]);
    const actualFormItems = props.mode === "simplified" && formItemDefinitions ? formItemDefinitions.mergedFields : data ? dataFormItems : formItems;
    const previousFormItems = React.useRef([]);
    React.useEffect(() => {
      if (!(rawData && rawData.isLoading)) {
        previousFormItems.current = actualFormItems != null ? actualFormItems : [];
      }
    }, [rawData, actualFormItems]);
    if (props.mode === "simplified" && rawData && "error" in rawData) {
      return /* @__PURE__ */ React.createElement("div", null, "Error when fetching data: ", rawData.error.message);
    }
    const childrenNode = props.mode === "simplified" ? /* @__PURE__ */ React.createElement(React.Fragment, null, (actualFormItems != null ? actualFormItems : []).map((formItem) => {
      var _a2, _b, _c;
      return /* @__PURE__ */ React.createElement(
        ErrorBoundary,
        {
          canvasEnvId: props["data-plasmic-canvas-envs"],
          message: `Error rendering input for ${(_b = (_a2 = formItem.label) != null ? _a2 : formItem.name) != null ? _b : "undefined"}`
        },
        /* @__PURE__ */ React.createElement(
          FormItemWrapper,
          __spreadProps(__spreadValues({}, omit(formItem, "key")), {
            noLabel: formItem.inputType === InputType.Checkbox || formItem.noLabel,
            valuePropName: (_c = formItem.valuePropName) != null ? _c : formItem.inputType === InputType.Checkbox ? "checked" : void 0,
            style: { width: "100%" }
          }),
          formItem.inputType === InputType.Text ? /* @__PURE__ */ React.createElement(Input, null) : formItem.inputType === InputType.Password ? /* @__PURE__ */ React.createElement(Input.Password, null) : formItem.inputType === InputType.TextArea ? /* @__PURE__ */ React.createElement(Input.TextArea, null) : formItem.inputType === InputType.Number ? /* @__PURE__ */ React.createElement(InputNumber, null) : formItem.inputType === InputType.Checkbox ? /* @__PURE__ */ React.createElement(AntdCheckbox, null, formItem.label) : formItem.inputType === InputType.Select ? /* @__PURE__ */ React.createElement(AntdSelect, { options: formItem.options }) : formItem.inputType === InputType.DatePicker ? /* @__PURE__ */ React.createElement(AntdDatePicker, { showTime: formItem.showTime }) : formItem.inputType === InputType.RadioGroup ? /* @__PURE__ */ React.createElement(
            AntdRadioGroup,
            {
              options: formItem.options,
              optionType: formItem.optionType,
              style: { padding: "8px" }
            }
          ) : null
        )
      );
    }), props.submitSlot) : props.children;
    const isSchemaForm = props.mode === "simplified" && !!props.data;
    const isLoadingData = rawData == null ? void 0 : rawData.isLoading;
    return /* @__PURE__ */ React.createElement(
      SchemaFormContext.Provider,
      {
        value: {
          mergedFields: formItemDefinitions == null ? void 0 : formItemDefinitions.mergedFields,
          minimalFullLengthFields: formItemDefinitions == null ? void 0 : formItemDefinitions.mergedFields,
          schema: rawData == null ? void 0 : rawData.schema
        }
      },
      /* @__PURE__ */ React.createElement(
        FormWrapper,
        __spreadProps(__spreadValues({
          key: remountKey
        }, rest), {
          children: childrenNode,
          formItems: rawData && rawData.isLoading ? previousFormItems.current : actualFormItems,
          ref: wrapperRef,
          style: isSchemaForm && isLoadingData ? {
            opacity: 0.5,
            transitionDelay: "250ms",
            transition: "1s"
          } : {}
        })
      ),
      isSchemaForm && isLoadingData && /* @__PURE__ */ React.createElement(
        "div",
        {
          style: {
            position: "absolute",
            width: "100%",
            height: "100%"
          }
        }
      )
    );
  }
);

export { SchemaForm as FormWrapper, SchemaForm, deriveFormFieldConfigs };
